feat: add support for all taggable Linode objects

wbollock opened the pull request:
This adds the ability to additionally tag volumes, nodebalanacers,
domains and lkeclusters. Report/JSON output has been updated to support
this new output too.

The readme changes highlight support for all the taggable Linode APIv4
objects, in conjunction with the new sample tagger config.

Now, the tagger config not only shows all the different resource types you
can tag but highlights some common use cases and syntax (multiple tags,
empty list).

wbollock committed:
feat: add support for all taggable Linode objects

This adds the ability to additionally tag volumes, nodebalanacers,
domains and lkeclusters. Report/JSON output has been updated to support
this new output too.

wbollock committed:
chore: update readme and example tagger config

The readme changes highlight support for all the taggable Linode APIv4
objects, in conjunction with the new sample tagger config.

The tagger config not only shows all the different resource types you
can tag but highlights some common use cases and syntax (multiple tags,
empty list).

tjhop commented on packaging/etc/tagger.yml:
```diff
@@ -6,3 +6,33 @@ tagger:
+      present:
+        - tagger_managed_1
+      absent: []
+  lkeclusters:
```
```suggestion
  lke_clusters:
```
Maybe?

This reads a bit odd to me. Open to opinions.

wbollock replied:
That's fine, I'm not biased either way. I was copying the API format of "LKEClusters" but it's more human readable with the underscore

tjhop commented on tagger.go:
```diff
@@ -41,7 +41,11 @@ type TagRule struct {
+	Volumes       []TagRule `yaml:"volumes"`
+	NodeBalancers []TagRule `yaml:"nodebalancers"`
+	Domains       []TagRule `yaml:"domains"`
+	LKEClusters   []TagRule `yaml:"lkeclusters"`
```
```suggestion
	LKEClusters   []TagRule `yaml:"lke_clusters"`
```
To support the config change

tjhop commented on tagger.go:
```diff
@@ -65,85 +69,240 @@ func newLinodeClient() linodego.Client {
+		tags = linode.Tags
+		label = linode.Label
+		linodeID = linode.ID
+	}
```
feedback:
- generally speaking, linodes VM instances are referred to as "linodes", so using the `linode` variable here to denote the type asserted linode object feels a bit confusing. maybe `s/linode/obj/` or somethig that more clearly indicates it can be a nodebalancer, volume, etc, at the switch statement and update references? 
- `s/tags/oldTags/`? 
- similar to the first point, these aren't all linode (instance) types anymore, so maybe `s/linodeID/id/`?

tjhop commented on tagger.go:
```diff
@@ -65,85 +69,240 @@ func newLinodeClient() linodego.Client {
-						// we can persist it through to the new tag set
-						newTags = append(newTags, tag)
-					}
+func compareTags(linodeObject interface{}, linodeTags []TagRule) (combinedNewTags []string, tags []string, linodeID int) {
```
The named return values read a bit odd here with the given names (especially when we explicitly provide the return values when we call anyway). Personally, I'd rather drop the named return values and just add a comment indicating that the function returns, in order, a slice of new tags, a slice of the old tags, and the object id.

tjhop replied:
```suggestion
func compareTags(linodeObject interface{}, tagRules []TagRule) (combinedNewTags []string, tags []string, linodeID int) {
```
Similarly here too, these aren't necessarily just linode compute instances, and I think this can benefit from an updated name to reflect that

tjhop commented on tagger.go:
```diff
@@ -65,85 +69,240 @@ func newLinodeClient() linodego.Client {
+	return combinedNewTags, tags, linodeID
+}
+
+func logTags(combinedNewTags []string, tags []string, linodeID int) {
+	if len(combinedNewTags) > 0 {
+		sort.Strings(combinedNewTags)
+		if !slices.Equal(tags, combinedNewTags) {
+			log.WithFields(log.Fields{
+				"linode_id": linodeID,
+				"old_tags":  tags,
+				"new_tags":  combinedNewTags,
+			}).Debug("Linode tag set updated")
+		}
+	}
+}
```
similar comments here about not necessarily being a compute instance type anymore. 

maybe we update the log fields to `s/linode_id/id` and add a new field that denotes the actual object type? ie, `vm`, `nodebalancer`, `domain`, etc? thoughts?

wbollock replied:
sgtm!

tjhop commented on tagger.go:
```diff
@@ -65,85 +69,240 @@ func newLinodeClient() linodego.Client {
-	if err != nil {
-		return err
-	}
+func updateLinodeInstanceTags(ctx context.Context, client linodego.Client, id int, tags *[]string, data string) error {
```
```suggestion
func updateObjectTags(ctx context.Context, client linodego.Client, id int, tags *[]string, data string) error {
```

I feel this can benefit from an updated name as well

tjhop commented on tagger.go:
```diff
@@ -65,85 +69,240 @@ func newLinodeClient() linodego.Client {
+			if !slices.Equal(updatedTags, *tags) {
+				return errors.New("call to update instance did not result in the expected tag set")
+			}
+		case "volumes":
+			updatedInstance, err := client.UpdateVolume(ctx, id, linodego.VolumeUpdateOptions{Tags: tags})
```
```suggestion
		case "volumes":
			updatedVolume, err := client.UpdateVolume(ctx, id, linodego.VolumeUpdateOptions{Tags: tags})
```
Perhaps I'm being too nit picky here, but this is another section where I feel it might read a bit clearer with a new name. I don't really care what, it only exists in this scope anyway. `obj`, `v`, whatever. just not referring to each of these inner scoped updated objects as instances sounds like a bit clearer to me

tjhop commented on tagger.go:
```diff
@@ -166,71 +325,169 @@ func sliceDifference(a, b []string) []string {
 	report := make(ReportMap)
+	var instanceType string
+	// iterate through every type of linode object and see what tags should change
+	for _, objects := range linodeObjects {
+		switch objects := objects.(type) {
+		case []linodego.Instance:
```
I think this reads much clearer 👍

tjhop commented on tagger.go:
```diff
@@ -340,13 +598,58 @@ func main() {
+	var domains []linodego.Domain
+	var lkeclusters []linodego.LKECluster
+
+	linodeObjects := []interface{}{
+		linodes,
+		volumes,
+		nodebalancers,
+		domains,
+		lkeclusters}
```
Looks like a `go fmt` is needed

tjhop commented on tagger.go:
```diff
@@ -340,13 +598,58 @@ func main() {
+		}
 	}
 
+	linodeObjects = []interface{}{
+		linodes,
+		volumes,
+		nodebalancers,
+		domains,
+		lkeclusters}
```
We redefine this?

tjhop commented on tagger.go:
```diff
@@ -340,13 +598,58 @@ func main() {
+		domains,
+		lkeclusters}
+
+	linodeTagSlice := make(map[string][]TagRule)
```
`linodeTagSlice`, but we're making a map 😛

tjhop reviewed (changes requested):
Overall it reads sane! Most of the changes requested are related to naming conventions and adjustments that I feel will make the code itself read a bit easier.

wbollock committed:
ref: change yaml config to refer to lke_clusters

wbollock committed:
ref: make most naming of linode/instance generic

In many variable names, function names and comments linode/instance was
used when really the data referred to a generic Linode API object of any
sort. This was changed to reflect the unknown nature of the linode
object.

wbollock committed:
ref: declare linodeObjects once and reinsert values

wbollock committed:
ref: improve error string for tag update failed

wbollock committed:
ref: add type field to logTags

Also cleanup a few Linode references

wbollock commented:
let me know if I should squash any of those commits together

wbollock committed:
ref: simpler fetching of linodeObjects

tjhop reviewed (approved):
Thanks @wbollock! I feel there's still some room for refinements, but things seem to be working well, so let's get the functionality in now and refine it as needed :+1:
